File: backend/internal/domain/gitprovider/gitprovider.go
package gitprovider

import (
	"time"
)

// Provider types (used by both user-level repository providers and repositories)
const (
	ProviderTypeGitHub = "github"
	ProviderTypeGitLab = "gitlab"
	ProviderTypeGitee  = "gitee"
	ProviderTypeSSH    = "ssh" // SSH-based Git server (no API)
)

// NOTE: Organization-level GitProvider has been removed.
// Git providers are now managed at the user level via:
// - UserRepositoryProvider (for importing repositories)
// - UserGitCredential (for Git operations)
// See: /backend/internal/domain/user/repository_provider.go
//      /backend/internal/domain/user/git_credential.go

// Repository represents a Git repository configured in the system
// Self-contained design: repository stores all necessary info, no git_provider_id dependency
type Repository struct {
	ID             int64 `gorm:"primaryKey" json:"id"`
	OrganizationID int64 `gorm:"not null;index" json:"organization_id"`

	// Provider info (self-contained, no foreign key to git_providers)
	ProviderType    string `gorm:"size:50;not null" json:"provider_type"`      // github, gitlab, gitee, generic
	ProviderBaseURL string `gorm:"size:255;not null" json:"provider_base_url"` // https://github.com, https://gitlab.company.com
	CloneURL        string `gorm:"size:500" json:"clone_url"`                  // Full clone URL

	ExternalID    string  `gorm:"size:255;not null" json:"external_id"`
	Name          string  `gorm:"size:255;not null" json:"name"`
	FullPath      string  `gorm:"size:500;not null" json:"full_path"`
	DefaultBranch string  `gorm:"size:100;default:'main'" json:"default_branch"`
	TicketPrefix  *string `gorm:"size:10" json:"ticket_prefix,omitempty"`

	// Visibility: "organization" (all members can see), "private" (only importer can see)
	Visibility       string `gorm:"size:20;not null;default:'organization'" json:"visibility"`
	ImportedByUserID *int64 `gorm:"index" json:"imported_by_user_id,omitempty"` // User who imported this repo

	IsActive bool `gorm:"not null;default:true" json:"is_active"`

	CreatedAt time.Time  `gorm:"not null;default:now()" json:"created_at"`
	UpdatedAt time.Time  `gorm:"not null;default:now()" json:"updated_at"`
	DeletedAt *time.Time `gorm:"index" json:"deleted_at,omitempty"` // Soft delete support
}

func (Repository) TableName() string {
	return "repositories"
}

File: web/src/lib/api/git-provider.ts
import { request } from "./base";

// Git Provider types
export interface GitProviderData {
  id: number;
  organization_id: number;
  provider_type: "github" | "gitlab" | "gitee" | "ssh";
  name: string;
  base_url: string;
  ssh_key_id?: number;
  is_default: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// Git Provider API
export const gitProviderApi = {
  list: () =>
    request<{ git_providers: GitProviderData[] }>("/api/v1/org/git-providers"),

  get: (id: number) =>
    request<{ git_provider: GitProviderData }>(`/api/v1/org/git-providers/${id}`),

  create: (data: {
    provider_type: string;
    name: string;
    base_url: string;
    client_id?: string;
    client_secret?: string;
    bot_token?: string;
    ssh_key_id?: number;
    is_default?: boolean;
  }) =>
    request<{ git_provider: GitProviderData }>("/api/v1/org/git-providers", {
      method: "POST",
      body: data,
    }),

  update: (id: number, data: {
    name?: string;
    base_url?: string;
    client_id?: string;
    client_secret?: string;
    bot_token?: string;
    ssh_key_id?: number;
    is_default?: boolean;
    is_active?: boolean;
  }) =>
    request<{ git_provider: GitProviderData }>(`/api/v1/org/git-providers/${id}`, {
      method: "PUT",
      body: data,
    }),

  delete: (id: number) =>
    request<{ message: string }>(`/api/v1/org/git-providers/${id}`, {
      method: "DELETE",
    }),

  testConnection: (id: number) =>
    request<{ success: boolean; message: string }>(`/api/v1/org/git-providers/${id}/test`, {
      method: "POST",
    }),
};
